staging/src/k8s.io/apiserver/pkg/authentication/token/cache/cached_token_authenticator.go

```diff
@@ -98,5 +110,17 @@ func (a *cachedTokenAuthenticator) AuthenticateToken(ctx context.Context, token
 }
 
 func keyFunc(auds []string, token string) string {
-	return fmt.Sprintf("%#v|%v", auds, token)
+	// individually hash each input to prevent ambiguities, i.e. "xy" + "z" == "x" + "yz"
+	audsHash := sha256.Sum256([]byte(fmt.Sprintf("%#v", auds)))
+	tokenHash := sha256.Sum256([]byte(fmt.Sprintf("%#v", token)))
+
+	// then hash all inputs with a random prefix
+	hash := sha256.New()
```

Comment:
```
hash := "crypto/hmac".New(sha256.New, randomCachePrefix)
```
Is almost what we are doing here and is a safer API than sha256. This looks fine now but a future refactor might introduce problems.

Reply:
Using HMAC doubled the allocations from `4` to `8` so I did not use this approach.

Reply:
This is vulnerable to length extension after the refactor, which partially undoes the dos prevention. Since this is subtle, i'm personally ok with making the gc work harder to prevent problems from creeping in. What does hmac look with a sync.Pool as @lavalamp suggested?

Reply:
This is vulnerable to length extension after the refactor, which partially undoes the dos prevention. Since this is subtle, i'm personally ok with making the gc work harder to prevent problems from creeping in. What does hmac look with a sync.Pool as @lavalamp suggested?

Reply:
Hash in a end marker or hash in the number of auds?